Add HTTP Gateway Protocol specification (#1119)
dfinity/portal: docs/references/http-gateway-protocol-spec.md

# The HTTP Gateway Protocol Specification

## Introduction

> The HTTP Gateway Protocol is an extension of the Internet Computer Protocol that allows conventional HTTP clients to interact with the Internet Computer network. This is important for software such as web browsers to be able to fetch and render client-side canister code, including HTML, CSS, and JavaScript as well as other static assets such as images or videos. The HTTP Gateway does this by translating between standard HTTP requests and requests that the Internet Computer Protocol will understand.

Reviewer: ```suggestion
The HTTP Gateway Protocol is an extension of the Internet Computer Protocol that allows conventional HTTP clients to interact with the Internet Computer network. This is important for software such as web browsers to be able to fetch and render client-side canister code, including HTML, CSS, and JavaScript as well as other static assets such as images or videos. The HTTP Gateway does this by translating between standard HTTP requests and [API canister calls](https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-interface) that the Internet Computer Protocol will understand.
```

> Such a gateway could be a stand-alone proxy, it could be implemented in web browsers (natively, via a plugin or a service worker) or in other ways. This document describes the interface and semantics of this protocol independent of a concrete Gateway so that all Gateway implementations can be compatible.

Reviewer: I would try to be consistent with capitalization:
```suggestion
Such a Gateway could be a stand-alone proxy, it could be implemented in web browsers (natively, via a plugin or a service worker) or in other ways. This document describes the interface and semantics of this protocol independent of a concrete Gateway so that all Gateway implementations can be compatible.
```

## Overview

> An HTTP request by an HTTP client is handled by these steps:

Reviewer: Just an idea, not sure it makes sense: Highlight who is doing the step by e.g., adding "Client", "Gateway", "Canister" to the step.

In addition, maybe we can add a step 0. that is "Intercept the HTTP request from the HTTP client" to have a counterpart for step 8.

Author: This makes sense, I'll add the "actor" in for each step

Author: Wait I'm confused, because you suggested to remove "the gateway" in the following steps :D could you clarify what you mean?

Reviewer: I suggested to remove the gateway in the following steps because the steps were not consistent. The first steps were written from the point of view of the gateway (e.g. "resolve the canister ID") and then they were written from the point of view as a bystander (e.g., "if applicable, the Gateway fetches"). I then thought we should make clear "who" is doing the step as step 4 doesn't really fit in with the rest. 

So I suggest, we either switch to something like 4: e.g., "The gateway resolves the canister ID..." or something like "Gateway: Resolve the canister ID..." and "Canister: Handle the request..."

Author: Understood, I'll update it

1. Resolve the canister ID that the request is intended for

> 2. Candid-encodes the HTTP request

Reviewer: Just for consistency:
```suggestion
2. Candid-encode the HTTP request
```

3. Invoke the canister via a query call to the `http_request` interface
4. The canister handles the request and returns an HTTP response, encoded in Candid, together with additional metadata

> 5. If requested by the canister, the Gateway sends the request again via an update call to `http_request_update`.

Reviewer: ```suggestion
5. If requested by the canister, send the request again via an update call to `http_request_update`.
```

> 6. If applicable, the Gateway fetches further body data via streaming query calls.

Reviewer: ```suggestion
6. If applicable, fetch further body data via streaming query calls.
```

> 7. If applicable, the Gateway validates the certificate of the response.

Reviewer: ```suggestion
7. If applicable, validate the certificate of the response.
```

> 8. The Gateway sends the response to the HTTP client.

Reviewer: ```suggestion
8. Send the response to the HTTP client.
```

## Canister ID Resolution

> The Gateway needs to know the canister id of the canister to talk to, and obtains that information from the hostname as follows:

Reviewer: ```suggestion
The Gateway needs to know the canister ID of the canister to talk to, and obtains that information from the hostname as follows:
```

1. If the hostname is in the following table, use the given canister ids:

> | Hostname             | Canister id                   |
> | -------------------- | ----------------------------- |
> | `identity.ic0.app`   | `rdmx6-jaaaa-aaaaa-aaadq-cai` |
> | `nns.ic0.app`        | `qoctq-giaaa-aaaaa-aaaea-cai` |
> | `dscvr.one`          | `h5aet-waaaa-aaaab-qaamq-cai` |
> | `dscvr.ic0.app`      | `h5aet-waaaa-aaaab-qaamq-cai` |
> | `personhood.ic0.app` | `g3wsl-eqaaa-aaaan-aaaaa-cai` |

Reviewer: Indentation
```suggestion
    | Hostname             | Canister id                   |
    | -------------------- | ----------------------------- |
    | `identity.ic0.app`   | `rdmx6-jaaaa-aaaaa-aaadq-cai` |
    | `nns.ic0.app`        | `qoctq-giaaa-aaaaa-aaaea-cai` |
    | `dscvr.one`          | `h5aet-waaaa-aaaab-qaamq-cai` |
    | `dscvr.ic0.app`      | `h5aet-waaaa-aaaab-qaamq-cai` |
    | `personhood.ic0.app` | `g3wsl-eqaaa-aaaan-aaaaa-cai` |
```

    method: text;
    url: text;
    headers: vec HeaderField;
    body: blob;
    certificate_version: opt nat;
};
```

The full [Candid](https://github.com/dfinity/candid/blob/master/spec/Candid.md) interface is described in [Canister HTTP Interface](#canister-http-interface).

The `certificate_version` field indicates the maximum supported version of [response verification](#response-verification). A value of `2` will request the current standard of [response verification](#response-verification), while a missing version or version 1 will request the [legacy standard](#legacy-response-verification).

- The method field contains the HTTP method in all upper case letters, e.g. `"GET"`.
- The URL field contains the URL from the HTTP request line, i.e. without protocol or hostname, and includes query parameters.
- The headers field contains the headers of the HTTP request.
- The body field contains the body of the HTTP request (without any content encodings processed by the Gateway).

## Query Calls

The encoded HTTP request is sent as a query call according to the [HTTPS Interface](https://internetcomputer.org/docs/current/references/ic-interface-spec#http-query).

> ## Response Decoding

Reviewer: Maybe to match the other section on HTTP Request Encoding

```suggestion
## HTTP Response Decoding
```

An HTTP response is decoded from the result of the [query call](#query-calls) using the following [Candid](https://github.com/dfinity/candid/blob/master/spec/Candid.md) interface:

```

> type HeaderField = record { text; text; };

Reviewer: Same as above?
```suggestion
type HeaderField = record { text: text; };
```

Notes:

- Not all HTTP Gateway implementations may be able to pass on all forms of headers. In particular, Service Workers are unable to pass on [forbidden headers](https://fetch.spec.whatwg.org/#forbidden-header-name).
- HTTP Gateways may add additional headers. In particular, the following headers may be set:
  - `access-control-allow-origin: \*`
  - `access-control-allow-methods: GET, POST, HEAD, OPTIONS`
  - `access-control-allow-headers: DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Cookie`
  - `access-control-expose-headers: Content-Length,Content-Range`
  - `x-cache-status: MISS`

## Response Verification

The HTTP Gateway will primarily be used to load static assets needed to run frontend canister code, so both low latency and security are essential for providing a good experience to end users. [Query calls](https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-query) are more performant but less secure than [Update calls](https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-call).

> Response verification fills the security gap left by query calls. It is a versioned subprotocol that allows for an HTTP Gateway to verify a certified response received as a result of performing a query call to the Internet Computer. Two versions are currently supported, the older version 1 only supports verification of the response body, whereas the newer version 2 supports verifying the full response, including headers and status code, and optionally the full request.

Reviewer: ```suggestion
Response verification fills the security gap left by query calls. It is a versioned subprotocol that allows for an HTTP Gateway to verify a certified response received as a result of performing a query call to the Internet Computer. Two versions are currently supported, the older version 1 only supports verification of the response body, whereas the newer version 2 supports verifying the full response, including headers, status code, and optionally the full request.
```

Response verification comes in two flavors, the current version is covered in this section and the legacy version is covered in [another section](#legacy-response-verification). The legacy version of response verification only includes a mapping of the request URL to the response body so it is quite limiting in what it can verify. The current version builds on the legacy version by optionally including the following extra parameters in the certification process:

- Request URL query params
- Request method
- Request headers
- Response status code
- Response headers

### Response Verification Outline

1. Case-insensitive search for the `IC-Certificate` response header.
   - If no such header is found, verification fails.
   - If the header value is not structured as per [the certificate header](#the-certificate-header), verification fails.
1. Parse the `certificate`, `tree`, and `expr_path` fields from the `IC-Certificate` header value as per [the certificate header](#the-certificate-header).

> 1. Case-insensitive search for the `IC-CertificationExpression` header.

Reviewer: Shouldn't this depend on the `version` in the step after the next?

Author: You're right, I'll reorder these, thanks

   - If the `version` field is equal to `2` then continue.
   - Otherwise, verification fails.
1. Validate that `expr_path` is the most specific expression path in the tree for the given request URL path.
   - If the absence of a more specific path cannot be proven (a lookup of more specific paths must return `Absent`) as per [lookup](https://internetcomputer.org/docs/current/references/ic-interface-spec/#lookup), verification fails.
1. Let `expr_hash` be the label of the node in the tree at path `expr_path`.
   - If no such label exists, verification fails.
   - If this node has sibling nodes, verification fails.
   - If `expr_hash` does not match the sha256 hash of the `IC-CertificateExpression` header value, verification fails.
   - If `no_certification` is set, verification succeeds.
   - Let `response_hash` be the response hash calculated according to [Response Hash Calculation](#response-hash-calculation)
   - If `no_request_certification` is set
     - If the `expr_hash` label node has an empty leaf node at the subpath `[response_hash]`, verification succeeds.

>      - Otherwise, the certification is invalid.

Reviewer: ```suggestion
     - Otherwise, verification fails.
```

   - Let `request_hash` be the request hash calculated according to [Request Hash Calculation](#request-hash-calculation)
     - If there is not an empty leaf node at the subpath `[request_hash, response_hash]`, verification fails.

### The Certificate Header

The `IC-Certificate` header is a structured header according to [RFC 8941](https://www.rfc-editor.org/rfc/rfc8941.html) with the following mandatory fields:

- `certificate`: [Base64 encoded](https://www.rfc-editor.org/rfc/rfc4648#section-4) string of self-describing, [CBOR-encoded](https://www.rfc-editor.org/rfc/rfc8949.html) bytes that decodes into a valid [certificate](https://internetcomputer.org/docs/current/references/ic-interface-spec/#certification).
- `tree`: [Base64 encoded](https://www.rfc-editor.org/rfc/rfc4648#section-4) string of self-describing, [CBOR-encoded](https://www.rfc-editor.org/rfc/rfc8949.html) bytes that decodes into a valid hash tree as per [certificate encoding](https://internetcomputer.org/docs/current/references/ic-interface-spec/#certification-encoding).

The following additional fields are mandatory for response verification version 2 and upwards:

- `version`

> - `expr_path`

Reviewer: Maybe we can add some info for the type: e.g., integer indicating the version, string containing the path or something like that.

Author: I agree this is missing, I'll add it, thanks

### Certificate Validation

> Certificate validation is performed as part of [response verification](#response-verification) as per [Canister Signatures](https://internetcomputer.org/docs/current/references/ic-interface-spec/#canister-signatures) and [Certification](https://internetcomputer.org/docs/current/references/ic-interface-spec/#certificate). It is expanded on here concerning [response verification](#response-verification)](#response-verification) for completeness:

Reviewer: ```suggestion
Certificate validation is performed as part of [response verification](#response-verification) as per [Canister Signatures](https://internetcomputer.org/docs/current/references/ic-interface-spec/#canister-signatures) and [Certification](https://internetcomputer.org/docs/current/references/ic-interface-spec/#certificate). It is expanded on here concerning [response verification](#response-verification) for completeness:
```

3. The decoded `certificate` must pass the following validations:
   - The certificate is signed by the root key or a subnet delegation signed by the root key.
   - If the certificate contains a subnet delegation, the delegation must be valid for the given canister.
   - The timestamp at the `/time` path must be recent, e.g. 5 minutes.
   - The subnet state tree in the certificate must reveal the canister's [certified data](https://internetcomputer.org/docs/current/references/ic-interface-spec/#system-api-certified-data).
4. The root hash of the decoded `tree` must match the canister's [certified data](https://internetcomputer.org/docs/current/references/ic-interface-spec/#system-api-certified-data).

### The Certificate Expression Header

The `IC-CertificateExpression` header carries additional information instructing the HTTP Gateway how to reconstruct the certification, it can instruct the HTTP Gateway to:

- Exclude the complete request/response pair or the request only
- Include specific request headers
- Include specific request URL query parameters
- Include or exclude specific response headers

> The format of the IC-CertificateExpression header is as follows:

Reviewer: Consistent with line 180:
```suggestion
The format of the `IC-CertificateExpression` header is as follows:
```

```
IC-CertificateExpression: default_certification(ValidationArgs{<literal field values>})
```

> The value of this header must have valid CEL syntax, such that `default_certification` could be implemented as a function provided by the HTTP Gateway to validate the certification.

Reviewer: Can we add a link to CEL syntax spec?

Author: Yes I'll add that, thanks

The properties supplied to this function are as follows:

- `certified_request_headers` - a list of request header names to include. This list can be empty.
  - Mutually exclusive with the `no_request_certification` property.
- `certified_query_parameters` - a list of request URL query parameter names to include. This list can be empty.
  - Mutually exclusive with the `no_request_certification` property.
- `certified_response_headers` - a list of response header names to include.
  - Must not include `IC-Certificate` or `IC-CertificateExpression`.

>   - Mutually exclusive with the `response_header_exlusions` property.

Reviewer: ```suggestion
  - Mutually exclusive with the `response_header_exclusions` property.
```

```

### Request Hash Calculation

The request hash is calculated as follows:

1. Let `request_headers_hash` be the [representation-independent hash](https://internetcomputer.org/docs/current/references/ic-interface-spec#hash-of-map) of the request headers:
   - Only include headers listed in the `certified_request_headers` field of [the certificate expression header](#the-certificate-expression-header).
     - If the field is empty or no value was supplied, no headers are included.
     - Headers can be repeated and each repetition should be included.
   - Include an additional `:ic-cert-method` header that contains the HTTP method of the request.
   - Include an additional `:ic-cert-query` header that contains a value according to the following steps:
     - Parse the query string and build a list of tuples `(<query_param_name>, <query_param_value>)` while maintaining the order.
     - Exclude all tuples where `<query_param_name>` does not exactly match a value listed in the `certified_query_parameters` field of [the certificate expression header](#the-certificate-expression-header). If `certified_query_parameters` is empty then the resulting list of tuples should also be empty.

>      - Concatenate each `<query_param_name>` with the corresponding `<query_param_value>` and then concatenate all of these concatenations using the original separators and order.

Reviewer: Why does the order matter?

Author: It's to make sure that any endpoints that happen to rely on the order can have different hashes for different orders. I'd generally consider it to be bad practice to rely on the order, but the idea was to be flexible in case someone wanted to do that for some reason...

Reviewer: Ok, got it. Can be resolved, I just don't have the rights to resolve it myself.

     - Calculate the sha256 hash of the UTF-8 representation of the resulting string.
2. Let `request_body_hash` be the sha256 of the request body.
3. Concatenate `request_headers_hash` and `request_body_hash` and calculate the sha256 of that concatenation.

### Response Hash Calculation

The response hash is calculated as follows:

1. Let `response_headers_hash` be the [representation-independent hash](https://internetcomputer.org/docs/current/references/ic-interface-spec#hash-of-map) of the response headers:
   - The `IC-Certificate` header is always excluded.
   - The `IC-CertificateExpression` header is always included.

>    - ## If the `certified_response_headers` field of [the certificate expression header](#the-certificate-expression-header) is present:

Reviewer: ```suggestion
   - If the `certified_response_headers` field of [the certificate expression header](#the-certificate-expression-header) is present:
```

Author: You're right... I'll fill this in, thanks!

   -
   - Headers can be repeated and each repetition should be included.
2. Let `response_body_hash` be the sha256 of the response body.

> 3. Concatenate `response_headers_haash` and `response_body_hash` and calculate the sha256 of that concatenation.

Reviewer: ```suggestion
3. Concatenate `response_headers_hash` and `response_body_hash` and calculate the sha256 of that concatenation.
```

The value of the `upgrade` field returned from `http_request_update` is ignored.

## Legacy Response Verification

Version 1 response verification only supports verifying a request path and response body pair with only one response per request path. This is quite restrictive in the number of scenarios it can support. For example, redirection or client-side caching is not safe since the status code and headers required to verify responses of that nature are not included in the certification. Upon a query call to a canister’s `http_request` method, a single malicious node or boundary node can modify these parts of the HTTP response, leading to the following issues:

- For any dApp, a single malicious node on the subnet can maliciously redirect the user anywhere they like.
- dApps cannot load the service worker when embedded within iFrames.

> - The use of redirects and cookies is unsafe as they can be manipulated by malicious nodes.

Reviewer: Isn't this the same as the first point. I would probably only keep this one.

Author: Agreed, thanks!

[Response Verification version 2](#response-verification) overcomes these issues.

The steps for response verification are as follows:

- See the [response verification outline](#response-verification-outline) for the full subprotocol description.
- The path `["http_assets", <url>]` exists in the `tree` and is a leaf with a value, where `<url>` is the utf8-encoded URL from the `HttpRequest`.
- Otherwise, the path `["http_assets", "/index.html"]` must exist in the `tree` and be a leaf.
- That leaf must contain the SHA-256 hash of the decoded body.
  - If the `streaming_strategy` field of the `HttpResponse` is set, all chunks are streamed and concatenated according to [response body streaming](#response-body-streaming) before decoding.
  - The body is decoded according to the `Content-Encoding` header if present.
  - Supported values for the `Content-Encoding` header include `gzip` and `deflate`.

## Canister HTTP Interface

The full [Candid](https://github.com/dfinity/candid/blob/master/spec/Candid.md) interface that a canister is expected to implement is as follows:

```

> type HeaderField = record { text; text; };

Reviewer: Again the same thing, but I really have no clue of candid:
```suggestion
type HeaderField = record { text: text; };
```

Author: `record { text: text; };` will create a record that has a single entry with the literal label `"text"` and a value of type `"text"`.
But `record { text; text}` creates a record with two entries that have auto-generated labels and values of type `"text"`, so it's the equivalent of writing `record { 0: text, 1: text }`: https://internetcomputer.org/docs/current/references/candid-ref#type-record--n--t--

Reviewer: Thanks for the explanation! Makes sense :)

type HttpRequest = record {
    method: text;
    url: text;
    headers: vec HeaderField;
    body: blob;
    certificate_version: opt nat;
};

type HttpResponse = record {
    status_code: nat16;
    headers: vec HeaderField;
    body: blob;
    upgrade : opt bool;
    streaming_strategy: opt StreamingStrategy;
};

// Each canister that uses the streaming feature gets to choose their concrete
// type; the HTTP Gateway will treat it as an opaque value that is only fed to
// the callback method

type StreamingToken = /* application-specific type */

type StreamingCallbackHttpResponse = record {
    body: blob;
    token: opt StreamingToken;
};

type StreamingStrategy = variant {
    Callback: record {
        callback: func (StreamingToken) -> (opt StreamingCallbackHttpResponse) query;
        token: StreamingToken;
    };
};

service : {
    http_request: (request: HttpRequest) -> (HttpResponse) query;
    http_request_update: (request: HttpRequest) -> (HttpResponse);
}
```

Not all of this interface is required. The following sections detail what can be optionally omitted depending on the requirements of the canister in question.

### Response Verification Interface

> The `certificate_version` field of the `HttpRequest` interface is optional depending on the version of [response verification](#response-verification) that the canister is implementing. It is omitted in older canisters that do not implement response verification version 2 or later.

Reviewer: I don't understand this paragraph. Doesn't the gateway set the `certificate_version` field in the request? The older canisters will just not know what to do with it.

Reviewer: Nevermind. I got it. The canister needs to implement the interface. Sorry!
